src/util/string.cpp

```diff
@@ -1068,14 +1068,33 @@ void safe_print_string(std::ostream &os, std::string_view str)
 	os.setf(flags);
 }
 
-
-v3f str_to_v3f(std::string_view str)
+std::optional<v3f> str_to_v3f(std::string_view str)
 {
-	v3f value;
-	Strfnd f(str);
-	f.next("(");
-	value.X = stof(f.next(","));
-	value.Y = stof(f.next(","));
-	value.Z = stof(f.next(")"));
-	return value;
+    str = trim(str);
+
+    if (str.empty()) {
+        return std::nullopt;
+    }
+
+    // Strip parentheses if they exist
+    if (str.front() == '(' && str.back() == ')') {
+        str.remove_prefix(1);
+        str.remove_suffix(1);
+    }
+
+    std::istringstream iss((std::string(str)));
+
+    const auto expect_delimiter = [&]() {
+        const auto c = iss.get();
+        return c == ' ' || c == ',';
+    };
+
+    v3f value;
+    if (!(iss >> value.X)) return std::nullopt;
+    if (!expect_delimiter()) return std::nullopt;
+    if (!(iss >> value.Y)) return std::nullopt;
+    if (!expect_delimiter()) return std::nullopt;
+    if (!(iss >> value.Z)) return std::nullopt;
+
```

Comment:
It's possible I forgot about this in my initial comment but we still need to check that there isn't a non-whitespace suffix here. So something like

```c++
while (!iss.eof()) if (!std::isspace(iss.get())) return std::nullopt;
```

please also consider adding a unit test like `1, 2, 3 test` which should result in `std::nullopt`